### DISCUSSION TITLE
New project repo template

### DISCUSSION THREAD

**Discussion by metazool** [NERC-CEH/rse_group#17] [Ideas]
Good suggestion from @matthewcoole and standup discussion about a minimum viable common template to bootstrap new repositories for research collaboration - what's good enough, less than cookiecutter, more than blank slate?

* LICENSE
* README
* test wrappings
* CI pipeline outline
* ...? 

https://book.the-turing-way.org/project-design/project-repo

**Comment by matthewcoole** [comment]
Perhaps to begin with we could just create an absolute minimum template maybe just (to discuss):
- LICENSE
- REAMDE
- Code of conduct?
- Contribution guidelines?

Then create further templates that are more "idealised" templates, these might contain all the things it would be great to include in a project repo. Things like:
- Testing hooks
- CI pipelines /w:
  - Test coverage reports
  - Documentation Sites
etc.

Then we can encourage people to initially use the minimum template and if they're more interested, incorporate additional elements from the idealised template.

@metazool I like the idea of holding to the Turing Way recommendations as a solid justification for what goes into the template(s).

**Comment by metazool** [comment]
https://github-pages.arc.ucl.ac.uk/python-tooling/ - I joined the UKRSE Slack community just to ask about common templates in the `beginner` channel, this links to what ARC at UCL do and also to templates specific to software ecologies

Has the heavy cookiecutter look to me though https://github.com/UCL-ARC/python-tooling#using-this-template

**Reply by metazool** [reply]
https://github.com/the-turing-way/reproducible-project-template - there's also this that implements the Turing guidelines but it looks like overkill to me, with a lot of context that doesn't belong in a repo. Useful reference point though

**Comment by longr** [comment]
It is probably worth looking at this paper (co-written by Sam Harrison in the Modelling team at CEH).  https://zenodo.org/records/4761867  I think it needs more levels.

**Comment by metazool** [comment]
If we are looking around for CEH prior art, @Kzra pointed me at this very nice R package https://github.com/NERC-CEH/jsdmstan/  - @fseaton doing development work on this

**Comment by metazool** [comment]
https://github.com/NERC-CEH/plankton_ml/ - we discussed this in standup when i made a PR for pipeline actions, minimal test harness, etc, but this was an attempt to express my view of minimum-viable.

The R project linked above is a nice example of existing research software maturity in the culture

**Comment by metazool** [comment]
Spotted this repo from @lewis-chambers that looks very like this discussion (from python POV) - minimal but complete-enough. What would we add just to adopt it, other than a default `LICENSE`?

https://github.com/NERC-CEH/python-template

To my eyes the `src` directory for the package looks like a Java-ism, would choose to omit it

**Reply by jmarshrossney** [reply]
To be fair I think there are some good reasons for using the `src/` layout. [This blog post](https://blog.ionelmc.ro/2014/05/25/python-packaging/#the-structure%3E) is quite old but I think still relevant, and the [pytest documentation](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#choosing-a-test-layout-import-rules) still "strongly" recommends it.

Not to say that WE MUST INSIST ON THE `src/` LAYOUT. Python packing is just like this it seems..

**Comment by jmarshrossney** [comment]
I agree with @matthewcoole in that I'd prefer that we started from minimal templates that are based on principles rather than tools, and then added a library of templates for different tool combinations, and things like documentation etc.

The most important thing in my mind is that we minimise lengthy and confusing conversations with colleagues about which tools/standards to adopt once a project is underway, so I think having a library of reference templates that we agree a priori are reasonable could be useful.

I definitly think it's futile to try to make one template to rule them all for Python projects.

**Comment by longr** [comment]
To throw this into the mix, instead of a template, what about a package to automate parts of this.   Users can pick and choose, and some meta functions could call groups of these based on the [level (See the paper co written by Sam Harrison)](https://zenodo.org/records/4761867) project.  Much like [__usethis__](https://usethis.r-lib.org/reference/index.html) in R which has various useful functions and will even setup CI/CD

**Comment by longr** [comment]
Random thought.  Along side this, (or instead?)  What about starting a review of tools and techniques?  I know there are a few out there, but nevwr really felt appropriate to research software, or compared minute differences in aspects not relevant.
